content/en/blog/_posts/2021-10-05-nsa-cisa-hardening.md

```diff
@@ -0,0 +1,413 @@
+movement to other worker or control plane nodes.
+
+Kubernetes 1.22 introduces
+an [alpha](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages) [feature](https://kubernetes.io/docs/tasks/administer-cluster/kubelet-in-userns/)
+that specifically reduces the impact of such a control plane component running
+as root user to a non-root user through user namespaces.
+
+Following container runtimes supports user namespaces / rootless mode
+
+1. [Docker Engine](https://docs.docker.com/engine/security/rootless/)
+2. [Podman](https://developers.redhat.com/blog/2020/09/25/rootless-containers-with-podman-the-basics)
+
+Some distributions that support running in rootless mode:
+
+1. [kind](https://kind.sigs.k8s.io/docs/user/rootless/)
+2. [k3s](https://rancher.com/docs/k3s/latest/en/advanced/#running-k3s-with-rootless-mode-experimental)
+3. [Usernetes](https://github.com/rootless-containers/usernetes)
+
+### Immutable container filesystems
+
+Guide highlights an often overlooked feature `readOnlyRootFileSystem` with a
+working example in Appendix B. This example limits execution and tampering of
+containers at runtime. Any read/write activity can then be limited to few
+directories by using `tmpfs` volume mounts.
+
+However, for some applications that modify container filesystem at runtime like 
+exploding a WAR or JAR file at container start up could face issues when 
+enabling this feature. To avoid this issue, it is important to consider making 
+minimal changes to filesystem at runtime when possible.
+
+### Building secure container images
+
+Guidance recommends running a scanner at deploy time as an admission controller,
+to prevent vulnerable or misconfigured pods from running in the cluster.
+Theoretically, this sounds like a good approach but there are several caveats to
+consider before this can be implemented in practice:
+
+- Depending on network bandwidth, available resources and scanner of choice,
+  scanning for vulnerabilities for an image can take an indeterminate amount of
+  time. This could lead to slower or unpredictable pod start up times, which
+  could result in spikes of unavailability when apps are serving peak load.
+- If the policy that allows or denies pod startup is made using incorrect or
+  incomplete data it could result in several false positive or false negative
+  outcomes like this:
+  - Openssl package inside a container image is detected as vulnerable. However,
+    the app uses golang native `crypto` package. Therefore, this vulnerability
+    is not in the code execution path and as such has minimal impact if it
+    remains unfixed
+  - A vulnerability is detected in openssl package for a debian base image.
+    However, the upstream debian community considers this as a Minor impact
+    vulnerability and as a result does not release a patch fix for this
+    vulnerability. Owner of this image is now stuck with a vulnerability that
```

Comment:
nit pick: `The owner of this image is now stuck with a vulnerability that cannot be fixed`